File: desktop/windows-app/src/GeregeDesktop.Client/Core/ApiClient.cs
// Gerege Template Platform V3.0
// Gerege Systems Development Team болон Claude AI хамтран бүтээв, 2026.

// BFF клиент.
//
// Апп нь Go backend-тэй ШУУД харьцахгүй — бүх хүсэлт платформын Next.js BFF-ээр
// явна (вэб, iOS, macOS-той яг ижил). Session нь httpOnly cookie
// (`dgov_access` / `dgov_refresh`): CookieContainer түүнийг хадгалж дараагийн
// хүсэлтэд өөрөө хавсаргана. Токен клиент код руу ХЭЗЭЭ Ч гарахгүй.
//
// BFF-ийн өөрчлөх route-ууд `x-dgov-csrf` header шаарддаг — native клиентэд
// Origin header байхгүй тул `checkOrigin` үүнийг л шалгана.

using System.Net;
using System.Text;
using System.Text.Json;
using GeregeDesktop.Client.Domain;

namespace GeregeDesktop.Client.Core;

internal sealed class ApiClient
{
    public static ApiClient Shared { get; } = new();

    /// <summary>
    /// Платформын BFF. `GEREGE_APP_URL` орчны хувьсагчаар дарж болно
    /// (локал `npm run dev` → http://localhost:3000).
    /// </summary>
    public static Uri BaseUrl { get; } = ResolveBaseUrl();

    private static Uri ResolveBaseUrl()
    {
        var raw = Environment.GetEnvironmentVariable("GEREGE_APP_URL");
        if (!string.IsNullOrWhiteSpace(raw)
            && Uri.TryCreate(raw.Trim(), UriKind.Absolute, out var parsed)
            && (parsed.Scheme == Uri.UriSchemeHttp || parsed.Scheme == Uri.UriSchemeHttps))
        {
            return parsed;
        }
        return new Uri("https://public.template.gerege.mn");
    }

    private readonly HttpClient _http;

    /// <summary>WebView2-оос хуулсан cookie энд хадгалагдана.</summary>
    public CookieContainer Cookies { get; } = new();

    private ApiClient()
    {
        var handler = new HttpClientHandler
        {
            CookieContainer = Cookies,
            UseCookies = true,
            AllowAutoRedirect = true,
        };
        _http = new HttpClient(handler) { BaseAddress = BaseUrl };
        _http.DefaultRequestHeaders.Add("Accept", "application/json");
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private async Task<(string Body, HttpStatusCode Status)> RequestAsync(
        string path, HttpMethod? method = null, object? body = null, CancellationToken ct = default)
    {
        method ??= HttpMethod.Get;
        using var req = new HttpRequestMessage(method, path);
        if (method != HttpMethod.Get)
        {
            req.Headers.Add("x-dgov-csrf", "1");
            var json = JsonSerializer.Serialize(body ?? new { });
            req.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var text = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        return (text, resp.StatusCode);
    }

    private static T? Unwrap<T>(string body)
    {
        var env = JsonSerializer.Deserialize<Envelope<T>>(body, JsonOptions);
        return env is null ? default : env.Data;
    }

    /// <summary>GET /api/me → хэрэглэгчийн профайл. Нэвтрээгүй бол null.</summary>
    public async Task<MeUser?> MeAsync(CancellationToken ct = default)
    {
        try
        {
            var (body, status) = await RequestAsync("/api/me", ct: ct).ConfigureAwait(false);
            if ((int)status >= 400) return null;
            return Unwrap<MeWrapper>(body)?.User;
        }
        catch (HttpRequestException)
        {
            return null;
        }
        catch (TaskCanceledException)
        {
            return null;
        }
    }

    /// <summary>
    /// eID нэгдсэн тоо. PKI_READ эрхгүй хэрэглэгчид 403 — энэ нь алдаа биш тул null.
    /// </summary>
    public async Task<EidSummary?> EidSummaryAsync(CancellationToken ct = default)
    {
        try
        {
            var (body, status) = await RequestAsync("/api/me/eid/summary", ct: ct).ConfigureAwait(false);
            if ((int)status >= 400) return null;
            return Unwrap<EidSummary>(body);
        }
        catch (HttpRequestException)
        {
            return null;
        }
        catch (TaskCanceledException)
        {
            return null;
        }
    }

    public async Task LogoutAsync(CancellationToken ct = default)
    {
        try
        {
            await RequestAsync("/api/auth/logout", HttpMethod.Post, new { }, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException)
        {
            // Сервер хүрэхгүй байсан ч локал cookie-г цэвэрлэнэ.
        }
        catch (TaskCanceledException)
        {
        }

        foreach (Cookie cookie in Cookies.GetCookies(BaseUrl))
        {
            cookie.Expired = true;
        }
    }

    /// <summary>WebView2-оос авсан cookie-г HttpClient-ийн сан руу хуулна.</summary>
    public void AdoptCookie(string name, string value, string domain, string path)
    {
        try
        {
            Cookies.Add(new Cookie(name, value, string.IsNullOrEmpty(path) ? "/" : path, domain));
        }
        catch (CookieException)
        {
            // Домэйн тохирохгүй cookie — алгасна (BFF-ийнх биш).
        }
    }
}

File: desktop/macos-app/Sources/Core/APIClient.swift
// Gerege Template Platform V3.0
// Gerege Systems Development Team болон Claude AI хамтран бүтээв, 2026.

// BFF клиент.
//
// Апп нь Go backend-тэй ШУУД харьцахгүй — бүх хүсэлт платформын Next.js BFF-ээр
// явна (вэб, iOS-той яг ижил). Session нь httpOnly cookie (`dgov_access` /
// `dgov_refresh`): URLSession түүнийг `HTTPCookieStorage.shared`-д хадгалж
// дараагийн хүсэлтэд өөрөө хавсаргана. Токен клиент код руу ХЭЗЭЭ Ч гарахгүй.
//
// BFF-ийн өөрчлөх (mutating) route-ууд `x-dgov-csrf` header шаарддаг —
// native клиентэд Origin header байхгүй тул `checkOrigin` үүнийг л шалгана.

import Foundation

enum APIError: Error, LocalizedError {
    case http(Int, String)
    case decoding
    case network(String)

    var errorDescription: String? {
        switch self {
        case .http(let code, let msg):
            return msg.isEmpty ? "Алдаа (\(code))." : "Алдаа (\(code)): \(msg)"
        case .decoding:
            return "Серверийн хариуг задлахад алдаа гарлаа."
        case .network(let msg):
            return "Сүлжээний алдаа: \(msg)"
        }
    }
}

final class APIClient {
    static let shared = APIClient()

    /// Платформын BFF. Энэ репогийн жишиг deployment.
    /// `GEREGE_APP_URL` орчны хувьсагчаар дарж болно (локал `npm run dev` → :3000).
    static let baseURL: URL = {
        if let raw = ProcessInfo.processInfo.environment["GEREGE_APP_URL"],
           let url = URL(string: raw.trimmingCharacters(in: .whitespaces)),
           url.scheme == "http" || url.scheme == "https" {
            return url
        }
        return URL(string: "https://open.gerege.mn")!
    }()

    private let session: URLSession

    private init() {
        let cfg = URLSessionConfiguration.default
        cfg.httpCookieStorage = HTTPCookieStorage.shared
        cfg.httpCookieAcceptPolicy = .always
        cfg.httpShouldSetCookies = true
        // Профайлын хариу кэшлэгдвэл гарсны дараа хуучин өгөгдөл харагдана.
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData
        session = URLSession(configuration: cfg)
    }

    // MARK: - Хүсэлт

    private func request(_ path: String, method: String = "GET",
                         body: [String: Any]? = nil) async throws -> (Data, HTTPURLResponse) {
        var req = URLRequest(url: Self.baseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if method != "GET" {
            req.setValue("1", forHTTPHeaderField: "x-dgov-csrf")
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body ?? [:])
        }
        do {
            let (data, resp) = try await session.data(for: req)
            guard let http = resp as? HTTPURLResponse else {
                throw APIError.network("хариу байхгүй")
            }
            return (data, http)
        } catch let err as APIError {
            throw err
        } catch {
            throw APIError.network(error.localizedDescription)
        }
    }

    /// `{ data: T }` дугтуйг задалж `T` буцаана.
    private func decodeData<T: Decodable>(_ data: Data, _ http: HTTPURLResponse) throws -> T {
        if http.statusCode >= 400 {
            let msg = (try? JSONDecoder().decode(Envelope<EmptyPayload>.self, from: data))?.message ?? ""
            throw APIError.http(http.statusCode, msg)
        }
        guard let env = try? JSONDecoder().decode(Envelope<T>.self, from: data),
              let payload = env.data else {
            throw APIError.decoding
        }
        return payload
    }

    private struct EmptyPayload: Decodable {}

    // MARK: - Профайл ба session

    /// Backend `/users/me` нь `data`-г `{ "user": {…} }` гэж боодог.
    private struct MeWrapper: Decodable { let user: MeUser }

    func me() async throws -> MeUser {
        let (data, http) = try await request("/api/me")
        let wrapped: MeWrapper = try decodeData(data, http)
        return wrapped.user
    }

    /// eID нэгдсэн тоо. PKI_READ эрхгүй хэрэглэгчид 403 — энэ нь алдаа биш тул `nil`.
    func eidSummary() async throws -> EidSummary? {
        let (data, http) = try await request("/api/me/eid/summary")
        if http.statusCode == 403 { return nil }
        return try? decodeData(data, http)
    }

    /// Session идэвхтэй эсэх — аппыг нээхэд шалгана.
    func isSignedIn() async -> Bool {
        (try? await me()) != nil
    }

    func logout() async {
        _ = try? await request("/api/auth/logout", method: "POST", body: [:])
        // BFF-ийн домэйны cookie-г локалаас цэвэрлэнэ.
        guard let host = Self.baseURL.host,
              let cookies = HTTPCookieStorage.shared.cookies else { return }
        for c in cookies where host.hasSuffix(c.domain) || c.domain.hasSuffix(host) {
            HTTPCookieStorage.shared.deleteCookie(c)
        }
    }
}
